File: py/trie.py
from typing import List, Self, Tuple, Set, Callable, Generator

import heapq
import numpy as np

import string
from dataclasses import dataclass, field

from datatypes import GridState

from util import (
    convert_matrix_to_letters,
    convert_letters_to_matrix,
    convert_list_to_string,
    string_to_alphabet_positions,
)


def default_letter_list_factory():
    return [None for _ in string.ascii_lowercase]


@dataclass(slots=True)
class TriePath:
    word: List[int]
    is_reverse: bool
    word_rank: float


@dataclass(slots=True)
class TrieNode:
    children: List[Self | None] = field(default_factory=default_letter_list_factory)
    path: TriePath = None

    def depth(self) -> int:
        child_depths = [child.depth() for child in self.children if child is not None]
        if len(child_depths) == 0:
            return 0
        return max(child_depths) + 1


# TODO: what if it's faster to just do a list of the words, at the size of our dictionary?
class WordTrie:
    """
    Provides functionality for getting valid letter states based on a slice.
    Supports both forwards and backwards embeddings of the provided word list.
    """

    def __init__(self, word_length_limits: Tuple[int, int] = (3, 10)):
        with open("word_lists/coca.txt") as f:
            word_list = f.readlines()

        smallest_word = 100
        biggest_word = 0

        filtered_words = []
        for word in word_list:
            word = word.strip().lower()
            if word_length_limits and (
                len(word) < word_length_limits[0] or len(word) > word_length_limits[1]
            ):
                continue
            smallest_word = min(smallest_word, len(word))
            biggest_word = max(biggest_word, len(word))
            filtered_words.append(word)
        self._word_limits = (smallest_word, biggest_word)
        self._num_entries = len(filtered_words)

        self.root = TrieNode()
        for word_placement, word in enumerate(filtered_words):
            word_rank = 1 - (word_placement / self._num_entries)
            word = string_to_alphabet_positions(word)
            self.add(word, word_rank)
            word.reverse()
            self.add(word, word_rank, reverse=True)

        self._depth = self.root.depth()

    def __len__(self) -> int:
        return self._num_entries

    def depth(self) -> int:
        return self._depth

    def word_limits(self) -> Tuple[int, int]:
        return self._word_limits

    def add(self, word: List[int], word_rank: float, reverse: bool = False):
        node = self.root
        for char in word:
            if char > 25:
                raise ValueError(
                    f"Recevied character '{convert_list_to_string([char])}' as part of word '{convert_list_to_string(word)}', which is unsupported!"
                )
            if node.children[char] is None:
                node.children[char] = TrieNode()
            node = node.children[char]
        if node.path is None or node.path.word_rank < word_rank:
            node.path = TriePath(word.copy(), reverse, word_rank)

    def get_valid_words(
        self, slice: List[int], blocked: List[bool], target_index: int
    ) -> Generator[Tuple[List[int], float, Tuple[int, int]], None, None]:
        max_children = len(self.root.children)
        slice_size = len(slice)
        minimum_word_length, maximum_word_length = self.word_limits()

        # Narrow in on only the part of the slice which is not blocked
        previous_blocked = -1
        next_blocked = slice_size
        for idx, is_blocked in enumerate(blocked):
            if is_blocked:
                if idx > target_index:
                    next_blocked = min(next_blocked, idx)
                if idx < target_index:
                    previous_blocked = max(previous_blocked, idx)

        # Find the possible window that our words might live in
        starting_index = max(previous_blocked + 1, target_index - maximum_word_length)
        ending_index = min(next_blocked, target_index + maximum_word_length)

        # If we are closer to the end of our valid word window, reverse it, so that we
        # don't have to check as many starting locations
        if target_index - starting_index > ending_index - target_index:
            # Reverse slice
            reversed_slice = slice.copy()
            if isinstance(reversed_slice, np.ndarray):
                reversed_slice = np.flip(reversed_slice)
            else:
                reversed_slice.reverse()
            # Reverse blocked
            reversed_blocked = blocked.copy()
            if isinstance(reversed_blocked, np.ndarray):
                reversed_blocked = np.flip(reversed_blocked)
            else:
                reversed_blocked.reverse()
            # Plug in the reversed inputs
            reversed_target = slice_size - 1 - target_index
            for word, value, (start, stop) in self.get_valid_words(
                slice=reversed_slice,
                blocked=reversed_blocked,
                target_index=reversed_target,
            ):
                word = word.copy()
                word.reverse()
                yield word, value, (slice_size - stop, slice_size - start)
            return

        # TODO: do not yield entries which include the blocked letter, but also don't include
        # entries which could include the blocked letter, in the future
        def _get_all_starting_at(starting_offset: int):
            tree_path = [
                None for _ in range(self.depth())
            ]  # Stack of (node, child_index)
            current_tree_index = 0
            current_node = self.root
            current_child_index = (
                0
                if slice[starting_offset] == GridState.OPEN
                else slice[starting_offset]
            )

            slice_index = starting_offset
            slice_entry = slice[slice_index]
            while True:
                # Check if we have exceeded the valid children of this node
                if (
                    current_child_index >= max_children
                    or slice_entry != GridState.OPEN
                    and slice_entry != current_child_index
                ):
                    if current_tree_index == 0:
                        return
                    current_tree_index -= 1
                    current_node, current_child_index = tree_path[current_tree_index]
                    tree_path[current_tree_index] = (
                        None  # TODO: hopefully remove this after debugging
                    )
                    current_child_index += 1
                    slice_index -= 1
                    slice_entry = slice[slice_index]
                    # Since we have updated our node, we want to restart the loop
                    continue

                # Check if the child at current_child_index exists and is valid
                if current_node.children[current_child_index] is not None and (
                    slice_entry == GridState.OPEN or slice_entry == current_child_index
                ):
                    next_node = current_node.children[current_child_index]

                    if next_node.path is not None:
                        if slice_index >= target_index:
                            yield next_node.path.word, next_node.path.word_rank, (
                                starting_offset,
                                slice_index + 1,
                            )

                    # If our next step will take us out of the slice boundaries,
                    # then we will not step forward, and instead continue to
                    # cycle through siblings of this node.
                    next_slice_index = slice_index + 1
                    if (
                        next_slice_index >= ending_index
                        or next_slice_index <= starting_index - 1
                    ):
                        current_child_index += 1
                    else:
                        # Append to our history
                        tree_path[current_tree_index] = (
                            current_node,
                            current_child_index,
                        )
                        current_tree_index += 1

                        # Get the next node
                        current_node = current_node.children[current_child_index]

                        # Get the next states, and figure out where to start our child index from
                        slice_index = next_slice_index
                        slice_entry = slice[slice_index]
                        current_child_index = (
                            0 if slice_entry == GridState.OPEN else slice_entry
                        )

                    # Since we have updated our node, we want to restart the loop
                    continue

                current_child_index += 1

        for starting_offset in range(
            starting_index,
            min(target_index, ending_index - minimum_word_length) + 1,
        ):
            yield from _get_all_starting_at(starting_offset)

    def get_top_k_valid_words(
        self,
        slice: List[int],
        blocked: List[bool],
        indices: Tuple[List[int], List[int]],
        target_index: str,
        scoring_function: Callable[[Tuple[List[int], List[int]], float], float] = None,
        k: int = 10,
        max_check: int = 2_000,
    ) -> List[List[int]] | None:
        k_largest = []

        random_values = np.random.random(max_check)

        idx = 0
        for word, word_rank, (start, stop) in self.get_valid_words(
            slice=slice, blocked=blocked, target_index=target_index
        ):
            word_indices = (
                indices[0][start:stop],
                indices[1][start:stop],
            )
            score = scoring_function(word_indices, word_rank, random_values[idx])

            if idx < k:
                heapq.heappush(k_largest, (score, idx, word, word_indices))
            elif score > k_largest[0][0]:
                heapq.heappushpop(k_largest, (score, idx, word, word_indices))

            if idx >= max_check - 1:
                break

            idx += 1

        return [
            (score, word, word_indices) for score, _, word, word_indices in k_largest
        ]  # If we want to sort later, we can

    def _words_from_slice(
        self, slice: List[int]
    ) -> Generator[Tuple[int, TrieNode], None, None]:
        current_nodes = [(0, self.root)]

        for current_index, entry in enumerate(slice):
            if entry != GridState.OPEN:
                # Traverse each of the current valid nodes
                current_nodes = [
                    (node_start, node.children[entry])
                    for node_start, node in current_nodes
                    if node.children[entry] is not None
                ]
            else:
                # Empty cells stop all current nodes
                current_nodes = []

            current_nodes.append((current_index + 1, self.root))

            # Check each of our current tries to see if we have a word
            # We are going in reverse order so that words that start earlier are
            # provided afterwards. Words which end later are also provided
            # afterwards. Both of these properties are useful to the `words`
            # function
            for start_position, node in reversed(current_nodes):
                if node.path is not None:
                    # We have found a word
                    yield start_position, node

    def coverage(self, slice: List[int]) -> np.ndarray:
        coverage = np.zeros(len(slice), dtype=bool)

        for start, node in self._words_from_slice(slice):
            coverage[start : start + len(node.path.word)] = True

        return coverage

    def words(self, slice: List[int], trim: bool = True) -> Set[str]:
        # Get all words in the given slice
        words = set()

        for _, node in self._words_from_slice(slice):
            new_word = node.path.word.copy()
            if node.path.is_reverse:
                new_word.reverse()
            new_word = convert_list_to_string(new_word)

            if trim:
                # Since we have gotten this word after any words which started
                # after this word and ended before (or right now), we can find
                # any words that are included in this one easily.
                words.difference_update({word for word in words if word in new_word})

            words.add(new_word)

        return words


if __name__ == "__main__":
    import cProfile
    import timeit

    trie = WordTrie()
    print(f"Created trie with depth {trie.root.depth()}")

    words = trie.words(slice=[])

    # def _score_word(
    #     word_indices: Tuple[List[int], List[int]],
    #     word_rank: int,
    #     random_value: float,
    # ) -> float:
    #     return word_rank + random_value

    # words = trie.get_valid_words(
    #     [GridState.OPEN, 2, 4, GridState.OPEN, GridState.OPEN, 8],
    #     blocked=[False, False, False, False, False, False],
    #     target_index=0,
    # )
    # print(list(words))

    # words = trie.get_top_k_valid_words(
    #     [GridState.OPEN for _ in range(40)],
    #     ([0 for _ in range(40)], [i for i in range(40)]),
    #     target_index=20,
    #     scoring_function=_score_word,
    # )
    # print(words)

    # cProfile.run(
    #     """words = trie.get_top_k_valid_words(
    #     [GridState.OPEN for _ in range(40)],
    #     ([0 for _ in range(40)], [i for i in range(40)]),
    #     target_index=20,
    #     scoring_function=_score_word,
    # )"""
    # )

    # print(
    #     "time per",
    #     timeit.timeit(
    #         """words = trie.get_top_k_valid_words(
    #     [GridState.OPEN for _ in range(40)],
    #     ([0 for _ in range(40)], [i for i in range(40)]),
    #     target_index=35,
    #     scoring_function=_score_word,
    # )""",
    #         globals={"trie": trie, "GridState": GridState, "_score_word": _score_word},
    #         number=20,
    #     )
    #     / 20,
    # )

File: src/trie.rs
use std::collections::HashSet;

use crate::consts::OPEN;
use crate::util::string_to_ints;

struct WordInfo {
    path: Vec<usize>,
    word: String,
    word_rank: f32,
}

const EMPTY: Option<Box<TrieNode>> = None;
const NUM_LETTERS: usize = 26;

const MIN_WORD: usize = 3;
const MAX_WORD: usize = 10;

pub struct ValidWordIterator<'a> {
    starting_index: usize,
    ending_index: usize,
    current_starting_index: usize,

    previous_nodes: Vec<(&'a TrieNode, usize)>,
    current_node: &'a TrieNode,
    current_node_child: usize,

    slice: &'a [usize],
    slice_position: usize,
}

impl<'a> ValidWordIterator<'a> {
    fn new(
        node: &'a TrieNode,
        slice: &'a [usize],
        starting_index: usize,
        ending_index: usize,
    ) -> Self {
        let current_child = if slice[starting_index] == OPEN {
            0
        } else {
            slice[starting_index]
        };

        println!("{}", starting_index);

        ValidWordIterator {
            starting_index: starting_index, // TODO: do we need this value?
            ending_index: ending_index,
            current_starting_index: starting_index,
            previous_nodes: Vec::with_capacity(NUM_LETTERS),
            current_node: node,
            current_node_child: current_child,
            slice: slice,
            slice_position: starting_index,
        }
    }
}

impl<'a> Iterator for ValidWordIterator<'a> {
    type Item = (Vec<usize>, f32, (usize, usize));

    fn next(&mut self) -> Option<Self::Item> {
        loop {
            // Check if we need to walk up the tree
            if self.current_node_child >= NUM_LETTERS
                || (self.slice[self.slice_position] != OPEN
                    && self.slice[self.slice_position] != self.current_node_child)
            {
                // Increment the current starting index, if necessary
                if self.previous_nodes.len() == 0 {
                    // Reset with the next starting index
                    self.current_starting_index += 1;
                    if self.current_starting_index == self.ending_index {
                        // Stop, we have found all valid words
                        return None;
                    }

                    // previous_nodes is already empty, so the current node will already be correct
                    self.slice_position = self.current_starting_index;
                    self.current_node_child = if self.slice[self.slice_position] == OPEN {
                        0
                    } else {
                        self.slice[self.slice_position]
                    };

                    // Restart the loop, so that we don't try to do the tree walk stuff
                    continue;
                }

                // Walk up the tree and step to next child
                (self.current_node, self.current_node_child) = self.previous_nodes.pop().unwrap();
                self.current_node_child += 1;

                // Step back in the slice
                self.slice_position -= 1;

                // Restart the loop
                continue;
            }

            // Check if there is a valid child at the current node child
            if (self.slice[self.slice_position] == OPEN
                || self.slice[self.slice_position] == self.current_node_child)
            {
                if let Some(next_node) = &self.current_node.children[self.current_node_child] {
                    let word_bounds = (self.current_starting_index, self.slice_position + 1);
                    // First, update our internal state, for the next time this function is called
                    if self.slice_position + 1 >= self.ending_index {
                        // Since we cannot step forward, we must increment the child
                        self.current_node_child += 1;
                    } else {
                        // Step forward
                        self.previous_nodes
                            .push((self.current_node, self.current_node_child));
                        self.current_node = next_node;

                        self.slice_position += 1;
                        self.current_node_child = if self.slice[self.slice_position] == OPEN {
                            0
                        } else {
                            self.slice[self.slice_position]
                        };
                    }

                    // Now, see if we have something to return
                    if let Some(word_info) = &next_node.word_info {
                        return Some((
                            word_info.path.clone(), // TODO: Beware, clone here
                            word_info.word_rank,
                            word_bounds,
                        ));
                    }
                }
            }

            // No valid child, check the next one
            self.current_node_child += 1;
        }
    }
}

pub struct TrieNode {
    children: [Option<Box<TrieNode>>; NUM_LETTERS],
    word_info: Option<WordInfo>,
}

impl TrieNode {
    fn new() -> Self {
        TrieNode {
            children: [EMPTY; NUM_LETTERS],
            word_info: None,
        }
    }

    pub fn build() -> Result<Self, String> {
        let mut root = TrieNode::new();
        // Load and filter words
        let words = include_str!("coca.txt");
        let filtered_words: Vec<&str> = words
            .split("\n")
            .map(|chunk| chunk.trim())
            .filter(|word| (word.len() > MIN_WORD) & (word.len() < MAX_WORD))
            .collect();
        // Add all of our words
        let num_words = filtered_words.len();
        for (rank_num, word) in filtered_words.iter().enumerate() {
            let word_rank = 1.0 - (rank_num as f32 / (num_words - 1) as f32);
            root.add(word.to_string(), word_rank)?;
        }
        Ok(root)
    }

    fn _add_one(&mut self, word_list: &Vec<usize>, word: &String, word_rank: f32) {
        let mut current_node = self;

        for character in word_list.iter() {
            let idx: usize = *character as usize;
            if current_node.children[idx].is_none() {
                current_node.children[idx] = Some(Box::new(TrieNode::new()));
            }
            current_node = current_node.children[idx].as_mut().unwrap();
        }
        current_node.word_info = Some(WordInfo {
            path: word_list.to_vec(),
            word: word.to_string(),
            word_rank: word_rank,
        })
    }

    fn add(&mut self, word: String, word_rank: f32) -> Result<(), String> {
        // Add the word in both directions
        let mut word_ints = string_to_ints(&word)?;
        self._add_one(&word_ints, &word, word_rank);
        word_ints.reverse();
        self._add_one(&word_ints, &word, word_rank);
        Ok(())
    }

    pub fn depth(self) -> u8 {
        let mut depth: u8 = 1;
        for potential_child in self.children {
            match potential_child {
                Some(child) => depth = depth.max(child.depth() + 1),
                None => {}
            }
        }
        depth
    }

    pub fn get_valid_words<'a>(
        &'a self,
        slice: &'a [usize],
        blocked: &'a [bool],
        target_index: usize,
    ) -> ValidWordIterator<'a> {
        debug_assert!(slice.len() == blocked.len());
        let slice_length = slice.len();

        let mut previous_blocked = None;
        let mut next_blocked = None;
        for (idx, is_blocked) in blocked.iter().enumerate() {
            if *is_blocked {
                if idx < target_index {
                    previous_blocked = Some(idx);
                } else if next_blocked.is_none() {
                    next_blocked = Some(idx);
                }
            }
        }

        let start = previous_blocked
            .map_or(0, |i| i + 1)
            .max(target_index.saturating_sub(MAX_WORD));
        let end = next_blocked
            .unwrap_or(slice_length)
            .min(target_index + MAX_WORD);

        // TODO: Add reversing logic

        ValidWordIterator::new(self, slice, start, end)
    }

    pub fn get_top_k_valid_words<'a>(
        &self,
        slice: &'a [usize],
        blocked: &'a [bool],
        indices: (&'a [usize], &'a [usize]),
        target_index: usize,
        scoring_function: fn((&[usize], &[usize]), f32, f32) -> f32,
        k: u16,
        max_words_to_check: u16,
    ) -> Vec<(f32, Vec<u8>, (&'a [usize], &'a [usize]))> {
        Vec::new()
    }

    // TODO: change slice input
    fn slice_words(self, slice: Vec<usize>) -> HashSet<String> {
        let mut current_nodes = vec![(0, &self)];
        let mut words: HashSet<String> = HashSet::new();

        for (index, slice_entry) in slice.iter().enumerate() {
            if *slice_entry < OPEN {
                // If we have a letter in this spot, then we need to get a new set of nodes
                let mut next_nodes = Vec::with_capacity(current_nodes.len());

                for node in &mut current_nodes {
                    if let Some(child) = node.1.children[*slice_entry as usize].as_deref() {
                        next_nodes.push((node.0, child));
                    }
                }

                current_nodes = next_nodes;
            } else {
                current_nodes = vec![];
            }

            current_nodes.push((index + 1, &self));

            // Now see if we have any words
            for (_, node) in current_nodes.iter().rev() {
                if let Some(word_info) = &node.word_info {
                    let word_to_add = word_info.word.to_string();

                    // First, remove from our set any words that are contained in this word
                    words.retain(|already_collected| !word_to_add.contains(already_collected));

                    // Add word to our set
                    words.insert(word_to_add);
                }
            }
        }

        words
    }

    // TODO: change slice input
    fn slice_coverage(self, slice: Vec<usize>) -> Vec<bool> {
        let mut current_nodes = vec![(0, &self)];
        let mut coverage: Vec<bool> = Vec::with_capacity(slice.len());

        for (index, slice_entry) in slice.iter().enumerate() {
            if *slice_entry < OPEN {
                // If we have a letter in this spot, then we need to get a new set of nodes
                let mut next_nodes = Vec::with_capacity(current_nodes.len());

                for node in &mut current_nodes {
                    if let Some(child) = node.1.children[*slice_entry as usize].as_deref() {
                        next_nodes.push((node.0, child));
                    }
                }

                current_nodes = next_nodes;
            } else {
                current_nodes = vec![];
            }

            current_nodes.push((index + 1, &self));

            // Now see if we have any words
            for (word_start, node) in current_nodes.iter().rev() {
                if let Some(word_info) = &node.word_info {
                    coverage[*word_start..*word_start + word_info.path.len()].fill(true);
                }
            }
        }

        coverage
    }
}
